### PR TITLE
feat: Initial implementation of many GCD API's using Swift Concurrency

### PR DESCRIPTION
**Pull request by scottmarchant**
# Summary

Initial implementation of common GCD API's using Swift Concurrency. The goal of this implementation is to enable wasm compilation for many repositories that currently use GCD.

Note this is a critical [part of larger effort](https://github.com/PassiveLogic/swift-web-examples/issues/1) to compile several open source Swift repositories to wasm

# Details

DispatchAsync is a temporary experimental repository aimed at implementing missing Dispatch support in the SwiftWasm toolchain.
Currently, [the Swift SDK for WebAssembly doesn't include Dispatch](https://book.swiftwasm.org/getting-started/porting.html#swift-foundation-and-dispatch). 
But it does support Swift Concurrency. DispatchAsync implements a number of common Dispatch API's using Swift Concurrency under the hood.

Refer to the README updates in this PR for more details.

### CODE REVIEW

File: Sources/DispatchAsync/DispatchTime.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+@available(macOS 13, *)
+extension DispatchTime {
+    public static func now() -> DispatchTime {
+        now
+    }
+
+    public var uptimeNanoseconds: UInt64 {
+        let beginning = uptimeBeginning
+        let rightNow = DispatchTime.now()
```
Comment (scottmarchant) [inline Sources/DispatchAsync/DispatchTime.swift:38]: Oops. I wrote a bad bug here. Just found it with some new test code I'm writing. For some reason I was thinking of this function as a static variable, but it's not, it's a member variable. So we should NOT be calling DispatchTime.now() in here.

File: Sources/DispatchAsync/DispatchQueue.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            targetQueue.async(execute: work)
+        } else {
+            if isMain {
+                Task { @MainActor [work] in
+                    DispatchQueue.$isMain.withValue(true) { @MainActor [work] in
+                        work()
+                    }
+                }
+            } else {
+                Task {
```
Comment (scottmarchant) [inline Sources/DispatchAsync/DispatchQueue.swift:80]: Found another bug with a new test I'm writing. This works if the attribute is .concurrent. But queues are supposed to be serial by default. And this is definitely NOT serial. 😅 

Will fix this as well.

File: Sources/DispatchAsync/DispatchGroup.swift [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+/// in Grand Central Dispatch. However, this class uses Swift Concurrency, instead of low-level threading API's.
+///
+/// The primary goal of this implementation is to enable WASM support for Dispatch.
+///
+/// Refer to documentation for the original [DispatchGroup](https://developer.apple.com/documentation/dispatch/dispatchgroup)
+/// for more details,
+@available(macOS 10.15, *)
+public class DispatchGroup: @unchecked Sendable {
+    /// Used to ensure FIFO access to the enter and leave calls
+    @globalActor
```
Comment (NeedleInAJayStack) [inline Sources/DispatchAsync/DispatchGroup.swift:27]: How does forcing this onto the global actor impact separately instantiated DispatchGroups? It seems like it would force all of them to execute sequentially, interleaving the results. i.e. Not only is this FIFO within the group, it looks like it's globally FIFO.

If this is the intended behavior, given that WASM is single-threaded, could we enforce or document that to avoid any expectation that this would operate in a multithreaded swift-concurrency environment?

File: Sources/DispatchAsync/DispatchTime.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        let uptimeDuration: Int64 = beginning.duration(to: rightNow).nanosecondsClamped
+        guard uptimeDuration >= 0 else {
+            assertionFailure("It shouldn't be possible to get a negative duration since uptimeBeginning.")
+            return 0
+        }
+        return UInt64(uptimeDuration)
+    }
+}
+
+// NOTE: The following was copied from swift-nio/Source/NIOCore/TimeAmount+Duration on June 27, 2025
```
Comment (NeedleInAJayStack) [inline Sources/DispatchAsync/DispatchTime.swift:48]: nit: A GitHub permalink might be a nicer way to reference the source of this code.

File: Sources/DispatchAsync/AsyncSemaphore.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift.org project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// Provides a semaphore implantation in `async` context, with a safe wait method. Provides easy safer replacement
+/// for DispatchSemaphore usage.
+@available(macOS 10.15, *)
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/AsyncSemaphore.swift:17]: I might keep this one as-is for now. I'll be needing to copy-paste this into some of the dependencies that need this code, because they won't accept adding a new formal dependency. Keeping the minimum OS requirement in the file makes that copy-paste scenario more manageable.

File: Sources/DispatchAsync/DispatchTime.swift (hunk 1 above)
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchTime.swift]: Fixed

File: Sources/DispatchAsync/DispatchTime.swift (hunk 4 above)
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchTime.swift:73]: Done

File: Sources/DispatchAsync/DispatchQueue.swift (hunk 2 above)
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchQueue.swift]: Done

File: Sources/DispatchAsync/DispatchSemaphore.swift [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        MainActor.assertIsolated()
+        assert(value > 0, "DispatchSemaphore is currently only designed for single-threaded use.")
+        value -= 1
+    }
+}
+
+#else
+
+@available(macOS 10.15, *)
+typealias DispatchSemaphore = AsyncSemaphore
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchSemaphore.swift]: I devised a special scheme that will be future proof to the possible future when Dispatch is available in the SwiftWasm toolchain. Non WASI platforms can use a special `@_spi(DispatchAsync) import DispatchAsync` to temporarily consume the code in place of Dispatch. That enables building and running this code in the unit test target against macOS, etc.

[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchSemaphore.swift]: Updated the readme with usage details.

**Review by NeedleInAJayStack — APPROVED** [review]

File: Sources/DispatchAsync/AsyncSemaphore.swift [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift.org project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// Provides a semaphore implantation in `async` context, with a safe wait method. Provides easy safe replacement
```
Comment (mpospese) [inline Sources/DispatchAsync/AsyncSemaphore.swift:15]: `implantation` typo?

File: Tests/DispatchAsyncTests/DispatchGroupTests.swift [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+@_spi(DispatchAsync) import DispatchAsync
+import Testing
+
+private typealias DispatchGroup = DispatchAsync.DispatchGroup
+
+@Test
+func dispatchGroupOrderCleanliness() async throws {
+    // Repeating this 100 times to help rule out
+    // edge cases that only show up some of the time
+    for index in 0 ..< 100 {
```
Comment (mpospese) [inline Tests/DispatchAsyncTests/DispatchGroupTests.swift:24]: What about replacing the `for` loop with parameterized testing?
e.g.
```
@Test("DispatchGroup Order Cleanliness", arguments: 0 ..< 100)
 func dispatchGroupOrderCleanliness(index: Int) async throws {
```

File: README.md [hunk 9]
```diff
@@ -1 +1,141 @@
...
+
+In the interim, to move wasm support forward, portions of DispatchAsync may be inlined (copy-pasted)
+into various libraries to enable wasm support. DispatchAsync is designed for this purpose, and has
+special `#if` handling to ensure that existing temporary usages will be elided without breakage
+the moment SwiftWasm adds support for `Dispatch` into the toolchain.
+
+# DispatchSemaphore Limitations
+
+The current implementation of `DispatchSemaphore` has some limitations. Blocking threads goes against the design goals of Swift Concurrency.
+The `wait` function on `DispatchSemaphore` goes against this goal. Furthermore, most wasm targets run on a single thread from the web
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply README.md:34]: DispatchSemaphore is implemented, but with a deprecation pointing people to use the async-friendly `AsyncSemaphore` instead.

**Review by JaapWijnen — APPROVED** [review]

File: Package.swift [hunk 10]
```diff
@@ -0,0 +1,22 @@
+// swift-tools-version: 6.0
+
+import PackageDescription
+
+let package = Package(
+    name: "dispatch-async",
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Package.swift]: Done

**Review by NeedleInAJayStack — APPROVED** [review]

File: Sources/DispatchAsync/AsyncSemaphore.swift [hunk 11]
```diff
@@ -0,0 +1,57 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/AsyncSemaphore.swift:3]: @clackary It is more of a hopeful statement to minimize the changes needed for swiftlang to adopt it.

[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/AsyncSemaphore.swift:3]: I essentially copied the headers from swift projects. The lint rules in the new shared swift CI github actions expect nearly this exact block I think.

File: Sources/DispatchAsync/DispatchTime.swift [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        ///
+        /// To keep this robust, we initialize `self.durationSinceBeginning`
+        /// to this value using a default value, which is guaranteed to run before any
+        /// initializers run. This guarantees that uptimeBeginning will be the very
+        /// first
+        @available(macOS 13, *)
+        private static let uptimeBeginning: ContinuousClock.Instant = ContinuousClock.Instant.now
+
+        /// See documentation for ``uptimeBeginning``. We intentionally
+        /// use this to guarantee a capture of `now` to uptimeBeginnin BEFORE
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchTime.swift]: Done

File: Sources/DispatchAsync/DispatchSemaphore.swift [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    /// Swift Concurrency api's with a safe async wait function.
+    #if !os(WASI)
+    @_spi(DispatchAsyncSingleThreadedSemaphore)
+    #endif
+    @available(
+        *,
+        deprecated,
+        renamed: "AsyncSemaphore",
+        message:
+            "DispatchSemaphore.wait is dangerous because of it's thread-blocking nature. Use AsyncSemaphore and Swift Concurrency instead."
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchSemaphore.swift:54]: Done

File: Sources/DispatchAsync/DispatchQueue.swift [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+public typealias DispatchQueue = DispatchAsync.DispatchQueue
+#endif
+
+extension DispatchAsync {
+    /// Drop-in replacement for ``Dispatch.DispatchQueue``, implemented using pure Swift.
+    ///
+    /// The primary goal of this implementation is to enable WASM support for Dispatch.
+    ///
+    /// Refer to documentation for the original [DispatchQueue](https://developer.apple.com/documentation/dispatch/dispatchqueue)
+    /// for more details,
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchQueue.swift]: Done

File: Sources/DispatchAsync/DispatchGroup.swift [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+extension DispatchAsync {
+    // MARK: - Public Interface for Non-Async Usage -
+
+    /// Drop-in replacement for ``Dispatch.DispatchGroup``, implemented using pure Swift.
+    ///
+    /// The primary goal of this implementation is to enable WASM support for Dispatch.
+    ///
+    /// Refer to documentation for the original [DispatchGroup](https://developer.apple.com/documentation/dispatch/dispatchgroup)
+    /// for more details,
```
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/DispatchGroup.swift]: Done

File: Sources/DispatchAsync/AsyncSemaphore.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (scottmarchant) [reply Sources/DispatchAsync/AsyncSemaphore.swift:3]: I'm probably going to merge this as is for now, but please @ mention me with your thoughts and I'll revise.
